### PR TITLE
checkout: use index checkout

### PR DESCRIPTION
**Pull request by efiop**
Needed for https://github.com/iterative/dvc/pull/9424

Part of #9333

Notable improvements:

* All file creation is parallelized now. Previously we did outputs and files within them step-by-step.
* We can capture and diff metadata that is not serializable into objects (e.g. isexec, link type, etc),
   which allows us to not waste our time relinking.

Other notes:

* Drops support for external outputs as a part of #9531
* The performance is hit and miss and needs a little more cleanup

### CODE REVIEW

**Comment by dberenbaum** [comment]
> * We can capture and diff metadata that is not serializable into objects (e.g. isexec, link type, etc),
>       which allows us to not waste our time relinking.

Does it fix https://github.com/iterative/dvc-data/issues/274?

**Comment by efiop** [comment]
@dberenbaum Not yet, no. It only handles isexec for now. The linking type detection is next step using the mechanism that is introduced here.

File: dvc/repo/checkout.py [hunk 1]
```diff
@@ -85,28 +90,71 @@ def checkout(
...
+        recursive=recursive,
+        with_deps=with_deps,
+        onerror=onerror,
+    )
+
+    old = build_data_index(view, self.root_dir, self.fs, compute_hash=True)
+    new = view.data["repo"]
+
+    for out in view.outs:
+        out.changed_cache()
```
Comment (efiop) [inline dvc/repo/checkout.py:116]: This is suboptimal. We need to just pass the index, collect oids in it and check all of them (probably parallelized).

**Comment by efiop** [comment]
Running with this to unblock https://github.com/iterative/dvc/pull/9424

**Pull request merged**